content/post/2023-10-20-cliff-notes-about-cli/index.Rmd

```diff
@@ -0,0 +1,216 @@
+
+```{r}
+thing <- "a string"
+cli::cli_li("Hey you provided this text: {thing}!")
+```
+
+Then, you can use classes in messages.
+Classes, like classes in CSS, help format the elements.
+For instance if `thing` is a variable, we write:
+
+```{r}
+thing <- "a string"
+cli::cli_li("Hey you provided this text: {thing} ({.var thing})!")
+```
+
+Curly braces can be nested if you are using a class and glue-like syntax at the same time:
+
+```{r}
+variable_name <- "blop"
+cli::cli_li("Hey you entered {.var {variable_name}}!")
+```
+
+See the [full list of classes](https://cli.r-lib.org/reference/inline-markup.html#classes) in cli docs.
+To mention only a few of them (you might notice clickability is a star here...if the terminal supports ANSI hyperlinks to runnable code (for instance, RStudio IDE)):
+
+-   `.run` means that the code in the message will be clickable! Best code hints ever!
+-   `.help` will have a clickable link to a help topic.
+-   `.file` will have a clickable link to a file.
+-   `.obj_type_friendly`, for instance `{.obj_type_friendly {mtcars}}`, prints an object in, well, a friendly way (thanks to Jon Harmon for reminding us about this one).
+
+It's well worth going through the list of classes at least once.
+
+### What if my string contains curly braces?
+
+If you want to actually communicate something with curly braces, you'll need to [double them to escape them](https://cli.r-lib.org/reference/inline-markup.html#escaping-and-).
+
+### What about plural?
+
+cli has support for [pluralization](https://cli.r-lib.org/reference/pluralization.html) (presumably only for 0/1/more than one, not for [more complex forms of pluralization](https://michaelchirico.github.io/potools/articles/translators.html#plurals)).
+
+### How to add a custom class / theme
+
+How to define a custom class or theme seems to be a bit under-documented at the moment, which is unsurprising as it's an advanced topic.
+Jenny Bryan [assumes](https://mastodon.social/@jennybryan@fosstodon.org/110675320334403080) adding a custom class needs to happen as part of a theme, which she does in [googlesheets4](https://github.com/tidyverse/googlesheets4/blob/fdb187643b324cd607f71cefa133bf49924f6e49/R/utils-ui.R#L1-L16).
+Michael McCarthy [recommends](https://mastodon.social/@mccarthymg@fosstodon.org/110673498809652527) looking at what cli itself does.
+
+### What about dark themes?
+
+Together, we cover both sides of having dark (Mo) and light (Maëlle) themes in RStudio.
+It would be easy to assume that this all would not work well on dark theme, but Mo can tell you that it all works seamlessly.
+The output colors change to be dark friendly!
+For that reason, user-overrides of colors can be a tricky thing, as you are no longer relying on the excellent tooling of cli to make sure this works in both light and dark mode.
+
+Open questions remain: How to let an user override it with their own?
```

Comment:
You can define your own themes, e.g. put this in your profile to make your variables underlined:
```r
options(cli.user_theme = list(.var = list("text-decoration" = "underline")))
cli::start_app() # not needed if the previous line is in the profile
cli::cli_text("Don't call a variable {.var variable}!")
```